refactor: inline `UndoWriteToDisk` and `WriteBlockToDisk` to reduce serialization calls

l0rinc opened the pull request:
`UndoWriteToDisk` and `WriteBlockToDisk` were delegating a subset of their functionality to single-use methods that didn't optimally capture a meaningful chunk of the algorithm, resulting in calculating things twice (serialized size, header size).
This change inlines the awkward methods (asserting that all previous behavior was retained), and in separate commits makes the usages less confusing.
Besides making the methods slightly more intuitive, the refactorings reduce duplicate calculations as well.

The speed difference is insignificant for now (~0.5% for the new `SaveBlockToDiskBench`), but are a cleanup for follow-ups such as https://github.com/bitcoin/bitcoin/pull/31539

andrewtoth reviewed (commented):
Concept ACK

Should we not prefer the more modern and explicit `uint32_t` vs `unsigned int`?

l0rinc commented:
> prefer the more modern and explicit uint32_t

I was hoping someone will recommend that - done: https://github.com/bitcoin/bitcoin/compare/e913e773926ecb72e327acf60c68655b5611cb7a..d69766164d177707ec7be19c4c188bd79ba3e4a3

ajtowns commented on src/node/blockstorage.cpp:
```diff
@@ -679,8 +679,8 @@ bool BlockManager::UndoWriteToDisk(const CBlockUndo& blockundo, FlatFilePos& pos
     // Write index header
-    unsigned int nSize = GetSerializeSize(blockundo);
-    fileout << GetParams().MessageStart() << nSize;
+    Assume(blockundo_size == GetSerializeSize(blockundo));
```
Calling `Assume(x)` always fully evaluates `x`, so there's no saving here... The only thing `Assume()` does differently in release mode is that it doesn't abort.

l0rinc replied:
Damn, you were right, thank you!
Pushed a separate commit that removes the `Assume`s (this way the reviewers and CI can make sure they're the same before/after) - plus found another debug log that has a similar problem, snuck it into this PR as well.

<details>
<summary>Test</summary>

> mkdir demo && cmake -B build -DCMAKE_BUILD_TYPE=Debug && cmake --build build -j$(nproc) && build/src/bitcoind -datadir=demo

```C++
static bool test()
{
    std::cout << "Assume was called!" << std::endl;
    return false;
}

static auto InitBlocksdirXorKey(const BlockManager::Options& opts)
{
    Assume(test());
...
```

prints in both cases, fails only with `Debug`

</details>

davidgumberg replied:
> Calling `Assume(x)` always fully evaluates `x`, so there's no saving here... The only thing `Assume()` does differently in release mode is that it doesn't abort.

I thought this wasn't the case if `x` can be determined by the compiler to have no side effects?

In any case https://github.com/bitcoin/bitcoin/issues/31178 shows that treating `assume(x)` as a no-op that can be placed in a hot path for free is a bad idea.

l0rinc replied:
Migrated the doc update to https://github.com/bitcoin/bitcoin/pull/31528

maflcko reviewed (changes requested):
Not sure about the changes.

What is the goal? Is this an optimization? If yes, how can it be observed?

Also, the changes seem to be based on a misunderstanding of the log and check macros.

l0rinc commented on src/blockencodings.cpp:
```diff
@@ -167,7 +167,9 @@ ReadStatus PartiallyDownloadedBlock::InitData(const CBlockHeaderAndShortTxIDs& c
-    LogDebug(BCLog::CMPCTBLOCK, "Initialized PartiallyDownloadedBlock for block %s using a cmpctblock of size %lu\n", cmpctblock.header.GetHash().ToString(), GetSerializeSize(cmpctblock));
+    if (LogAcceptCategory(BCLog::CMPCTBLOCK, BCLog::Level::Debug)) {
+        LogDebug(BCLog::CMPCTBLOCK, "Initialized PartiallyDownloadedBlock for block %s using a cmpctblock of size %lu\n", cmpctblock.header.GetHash().ToString(), GetSerializeSize(cmpctblock));
+    }
```
> // Use a macro instead of a function for conditional logging to prevent
> // evaluating arguments when logging for the category is not enabled.

Thanks, dropped the logger related commit. What's the reason for the logger avoiding evaluation but `Assume` not doing it?

> What is the goal? Is this an optimization?

As stated in the description `a costly calculation, we can deduplicate it`. I'll clarify further in the description.

> If yes, how can it be observed?

A quick peek can be seen in https://github.com/bitcoin-dev-tools/benchcoin/pull/66, but I'm currently running several full reindex-chainstates to see how much effect it has - previous measurements show 2-3%.

ryanofsky commented on src/node/blockstorage.h:
```diff
@@ -175,7 +175,7 @@ class BlockManager
      * (BLOCK_SERIALIZATION_HEADER_SIZE)
      */
-    bool WriteBlockToDisk(const CBlock& block, FlatFilePos& pos) const;
+    bool WriteBlockToDisk(const CBlock& block, uint32_t block_size, FlatFilePos& pos) const;
```
In commit "refactor: Cache block serialized size for consecutive calls" (0cb15046b039d0e6efd2b57075bade46e1a10237)

Would be good to document new `WriteBlockToDisk` `block_size` and `UndoWriteToDisk` `blockundo_size` arguments to clarify how they relate to the `pos` argument. Specifically it would be good to say that size arguments should be set to the size of data without the header, but the `pos` argument will increase by size of the data plus the header.

l0rinc replied:
I tried doing that, but noticed that all variants were awkward.
Realized that the code smell is caused by extracting a part of the write methods that isn't a meaningful subset of the functionality: it's tied closely to the only caller (needs the header size twice, recalculated undo serializes size, returns multiple branches, modifies parameter, needs documentation).
So I've recreated the PR completely, inlining these single-use methods, which also solves the serialization size calculations more naturally.

ryanofsky reviewed (approved):
Code review ACK 9a7e1ced7c3bb17193c7401181365c4075d45ec2. This seems like a safer, more efficient approach, but I still think the API is very confusing.

No need to address here, but for a followup I think it would make more sense for  WriteBlockToDisk instead of SaveBlockToDisk call FindNextBlockPos, and for UndoWriteToDisk instead of WriteUndoDataForBlock to be call FindUndoPos, so higher level SaveBlockToDisk / WriteUndoDataForBlock functions don't contain any logic dealing with header fields.

I also find names of these functions to be inconsistent, overlong and confusing. Would rename:

- SaveBlockToDisk to SaveBlock
- WriteBlockToDisk to WriteBlock
- WriteUndoDataForBlock to SaveBlockUndo
- UndoWriteToDisk to WriteBlockUndo

l0rinc committed:
refactor,blocks: deduplicate block's serialized size calculations

For consistency `UNDO_DATA_DISK_OVERHEAD` was also extracted to avoid the constant's ambiguity.
Asserts were added to help with the review - they are removed in the next commit.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

l0rinc committed:
refactor,blocks: inline `UndoWriteToDisk`

`UndoWriteToDisk` wasn't really extracting a meaningful subset of the `WriteUndoDataForBlock` functionality, it's tied closely to the only caller (needs the header size twice, recalculated undo serializes size, returns multiple branches, modifies parameter, needs documentation).

The inlined code should only differ in these parts (modernization will be done in other commits):
* renamed `_pos` to `pos` in `WriteUndoDataForBlock` to match the parameter name;
* inlined `hashBlock` parameter usage into `hasher << block.pprev->GetBlockHash()`;
* changed `return false` to `return FatalError`;
* capitalize comment.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

l0rinc committed:
refactor,blocks: inline `WriteBlockToDisk`

Similarly, `WriteBlockToDisk` wasn't really extracting a meaningful subset of the `SaveBlockToDisk` functionality, it's tied closely to the only caller (needs the header size twice, recalculated block serializes size, returns multiple branches, mutates parameter).

The inlined code should only differ in these parts (modernization will be done in other commits):
* renamed `blockPos` to `pos` in `SaveBlockToDisk` to match the parameter name;
* changed `return false` to `return FlatFilePos()`.

Also removed remaining references to `SaveBlockToDisk`.

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>

l0rinc committed:
refactor,blocks: remove costly asserts and modernize affected logs

When the behavior was changes in a previous commit (caching `GetSerializeSize` and avoiding `AutoFile.tell`), (static)asserts were added to make sure the behavior was kept - to make sure reviewers and CI validates it.
We can safely remove them now.

Logs were also slightly modernized since they were trivial to do.

Co-authored-by: Anthony Towns <aj@erisian.com.au>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

maflcko reviewed (commented):
concept ack, if this improves a benchmark. In the future, it would be good to mention a speedup in the pull request description, so that reviewers can see the motivation and goal for the change.

l0rinc commented on src/util/check.h:
```diff
@@ -93,6 +93,7 @@ constexpr T&& inline_assertion_check(LIFETIMEBOUND T&& val, [[maybe_unused]] con
  * - For fatal errors, use Assert().
  * - For non-fatal errors in interactive sessions (e.g. RPC or command line
  *   interfaces), CHECK_NONFATAL() might be more appropriate.
+ *   Note that the assumption is always fully evaluated - even in non-debug builds.
```
Since `Assume` is a macro, it could have been written differently to avoid parameter evaluation (similarly to `LogDebug`), so I think it needs a comment.
But I'm fine doing it in a different PR is this is remotely controversial.

l0rinc replied:
Migrated it to https://github.com/bitcoin/bitcoin/pull/31528

l0rinc replied:
Thank you for clarifying.
But given that we've bumped into this multiple times, do you think https://github.com/bitcoin/bitcoin/pull/31528 isn't a valid change?

BrandonOdiwuor reviewed (commented):
Concept ACK

theuni commented:
> concept ack, if this improves a benchmark. In the future, it would be good to mention a speedup in the pull request description, so that reviewers can see the motivation and goal for the change.

Agree with this. If this actually shows up as significant in real workloads (IBD), concept ACK and we could potentially take this further by caching the size even earlier (as part of deserializing over the wire) to avoid the need for the calculation at all.

But if there's no noticeable speedup, I'm not a fan of muddying the api.

l0rinc commented:
> If this actually shows up as significant in real workloads

Thanks for the reviews, I'm running full IBD benchmarks currently, we'll see the results shortly (can't just do a quick reindex-chainstate).

I have two other changes in queue that will be based on this refactor (including https://github.com/bitcoin/bitcoin/pull/31144 which I've drafted until these are sorted).
The 3 changes together seem to result in >5% speedup for IBD (every kind, regardless of dbcache or prunedness) - but those benchmarks are also still running.

l0rinc commented:
> we could potentially take this further by caching the size even earlier

Absolutely, but that's a bigger change (would cache the serialized sizes in `CBlock`, guarding against any other mutation (which requires better encapsulation), storing `GetSerializeSize` for `TX_NO_WITNESS()` and `TX_WITH_WITNESS()` lazily, similarly to the existing `checked*` flags)... but that's a big change, affecting a lot of consensus code, I'm still working on that and will push in a separate PR.

l0rinc commented:
@maflcko, @theuni, the reindex and IBD didn't show any speedup, so I've added a `SaveBlockToDiskBench` microbenchmark which only revealed a tiny speedup in itself, so I've demoted this from an optimization to a refactoring.
This PR is now meant as a cleanup for follow-ups such as https://github.com/bitcoin/bitcoin/pull/31539 and https://github.com/bitcoin/bitcoin/pull/31144 where the total IBD speedups (until 870k blocks) add up to 9% - and `ReadBlockFromDiskTest` benchmark is 2x faster.

<details>
<summary>Details</summary>

```bash
hyperfine \
--runs 2 \
--export-json /mnt/my_storage/ibd-xor-buffered.json \
--parameter-list COMMIT b042c4f0538c6f9cdf8efbcef552796851e38a85,8be7a6f07b1887103b7613e3f913ae708adcb42f \
--prepare 'rm -rf /mnt/my_storage/BitcoinData/* && git checkout {COMMIT} && git clean -fxd && git reset --hard && cmake -B build -DCMAKE_BUILD_TYPE=Release -DBUILD_UTIL=OFF -DBUILD_TX=OFF -DBUILD_TESTS=OFF -DENABLE_WALLET=OFF -DINSTALL_MAN=OFF && cmake --build build -j$(nproc)' \
'COMMIT={COMMIT} ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=870000 -printtoconsole=0'

Benchmark 1: COMMIT=b042c4f0538c6f9cdf8efbcef552796851e38a85 ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=870000 -printtoconsole=0
  Time (mean ± σ):     39790.800 s ± 26.465 s    [User: 51319.040 s, System: 3420.884 s]
  Range (min … max):   39772.086 s … 39809.514 s    2 runs

Benchmark 2: COMMIT=8be7a6f07b1887103b7613e3f913ae708adcb42f ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=870000 -printtoconsole=0
  Time (mean ± σ):     36396.906 s ± 27.299 s    [User: 47531.210 s, System: 3384.100 s]
  Range (min … max):   36377.603 s … 36416.210 s    2 runs

Summary
  COMMIT=8be7a6f07b1887103b7613e3f913ae708adcb42f ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=870000 -printtoconsole=0 ran
    1.09 ± 0.00 times faster than COMMIT=b042c4f0538c6f9cdf8efbcef552796851e38a85 ./build/src/bitcoind -datadir=/mnt/my_storage/BitcoinData -stopatheight=870000 -printtoconsole=0
```

</details>

I wanted to split this from https://github.com/bitcoin/bitcoin/pull/31144 to reduce risk, but if you think it's easier to review them in a single PR, I don't mind.

TheCharlatan commented on src/index/coinstatsindex.cpp:
```diff
@@ -123,7 +123,7 @@ bool CoinStatsIndex::CustomAppend(const interfaces::BlockInfo& block)
         // will be removed in upcoming commit
         const CBlockIndex* pindex = WITH_LOCK(cs_main, return m_chainstate->m_blockman.LookupBlockIndex(block.hash));
-        if (!m_chainstate->m_blockman.UndoReadFromDisk(block_undo, *pindex)) {
+        if (!m_chainstate->m_blockman.UndoRead(block_undo, *pindex)) {
```
In commit 0dd386994bafd67081c4521673018ddd22e2d63c

Can we call this `ReadBlockUndo` instead?

l0rinc replied:
That was inconsistent, indeed - [fixed](https://github.com/bitcoin/bitcoin/compare/0dd386994bafd67081c4521673018ddd22e2d63c..2ff0ea366c61b2fcb80ad1f711480915c7a7aa2e)!

hodlinator reviewed (commented):
Concept ACK cba4378072a6bd43f3e37a7d5e662d3041681566

Removes repetitive calls to `GetSerializeSize()`. And avoids calling `.tell()`.

Nothing blocking.


### PR summary

Repetitive mention of #31539.


### Commit history

Tested splitting up 96c9b578a6b85fac9977cdf25bf2e52d04645bd4 into one which refactors existing code (225d294070fc403e03028b9fab714bd9dc4f3307), and a second commit (61a7254e1b3d4b5159ffcab6e1096da32dc1ace0) which renames the file and adds `SaveBlockBench()`. Keeping the changes small in the rename-commit means that Git clients correctly show it as a rename instead of add+delete.

#### Current PR

```
₿ git -c log.showSignature=false log --oneline --follow readwriteblock.cpp
2ff0ea366c (HEAD -> pr/31490) scripted-diff: rename block and undo functions for consistency
96c9b578a6 bench: add SaveBlockBench
```

#### With suggested splitting of commit 96c9b578a6b85fac9977cdf25bf2e52d04645bd4

```
₿ git -c log.showSignature=false log --oneline --follow readwriteblock.cpp
9eba6eae8b (HEAD -> pr/31490_suggestions, myfork/pr/31490_suggestions) scripted-diff: rename block and undo functions for consistency
61a7254e1b bench: add SaveBlockBench
225d294070 refactor,bench: Stop calling benchmarks test, etc
faecca9a85 test: Use span for raw data
fab0e834b8 bench: [refactor] iwyu
d9e477c4dc validation, blockstorage: Separate code paths for reindex and saving new blocks
1c4b9cbe90 bench: add readblock benchmark
```

ryanofsky reviewed (approved):
Code review ACK 2ff0ea366c61b2fcb80ad1f711480915c7a7aa2e. New version of this PR is different and I think easier to follow than the previous version. All the changes now seem like obvious code cleanups.

Might consider renaming PR from "cache block[undo] serialized size" because I'm not sure it's even accurate to call not pointlessly recomputing the same values twice in a row "caching".

l0rinc committed:
refactor,bench: rename bench/readblock.cpp to bench/readwriteblock.cpp

Done in separate commit to simplify review.
Also renames benchmarks, since they're not strictly tests.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

l0rinc committed:
bench: add SaveBlockBench

l0rinc committed:
scripted-diff: rename block and undo functions for consistency

Co-authored-by: Ryan Ofsky <ryan@ofsky.org>
Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

-BEGIN VERIFY SCRIPT-
grep -r -wE 'WriteBlock|ReadRawBlock|ReadBlock|WriteBlockUndo|ReadBlockUndo' $(git ls-files src/ ':!src/leveldb') && \
    echo "Error: One or more target names already exist!" && exit 1
sed -i \
    -e 's/\bSaveBlockToDisk/WriteBlock/g' \
    -e 's/\bReadRawBlockFromDisk/ReadRawBlock/g' \
    -e 's/\bReadBlockFromDisk/ReadBlock/g' \
    -e 's/\bWriteUndoDataForBlock/WriteBlockUndo/g' \
    -e 's/\bUndoReadFromDisk/ReadBlockUndo/g' \
    $(git ls-files src/ ':!src/leveldb')
-END VERIFY SCRIPT-

l0rinc commented on src/bench/readwriteblock.cpp:
```diff
@@ -0,0 +1,68 @@
+#include <memory>
+#include <vector>
+
+CBlock CreateTestBlock()
```
Done

l0rinc commented on src/node/blockstorage.h:
```diff
@@ -330,7 +324,7 @@ class BlockManager
     CBlockFileInfo* GetBlockFileInfo(size_t n);
 
-    bool WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex& block)
+    bool SaveBlockUndo(const CBlockUndo& blockundo, BlockValidationState& state, CBlockIndex& block)
```
I've updated the scripted diff to use consistent Read/Write terminology - dropping the `[To/From]Disk` part which also just seems like noise to me. Thanks for the proposals.

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -945,7 +945,9 @@ bool BlockManager::WriteUndoDataForBlock(const CBlockUndo& blockundo, BlockValid
     if (block.GetUndoPos().IsNull()) {
         FlatFilePos pos;
-        if (!FindUndoPos(state, block.nFile, pos, ::GetSerializeSize(blockundo) + 40)) {
+        const uint32_t blockundo_size{static_cast<uint32_t>(GetSerializeSize(blockundo))};
```
I usually went in the other direction to make sure we know the sizes exactly.
Since `size_t` doesn't really serialize (e.g. in `fileout << GetParams().MessageStart() << blockundo_size` -> `call to 'Serialize' is ambiguous`) - but I made it into a `unsigned int` (I don't particularly like it, but it's not that important).

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/31490#discussion_r1908760710

> I made it into a `unsigned int` (I don't particularly like it, but it's not that important).

Thanks, your change makes sense and to be clear I don't like code writing `unsigned int` everywhere either, even though I do think `unsigned int` is a reasonable type for this code to be using. I think more ideally src/flatfile.h would define something like:

```c++
using FilePos = unsigned int;
```

 and then code could reference the `FilePos` type instead of unsigned int.

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -1092,33 +1092,24 @@ bool BlockManager::ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatF
-    // defined as BLOCK_SERIALIZATION_HEADER_SIZE)
-    nBlockSize += static_cast<unsigned int>(BLOCK_SERIALIZATION_HEADER_SIZE);
-    FlatFilePos pos{FindNextBlockPos(nBlockSize, nHeight, block.GetBlockTime())};
+    const uint32_t block_size{static_cast<uint32_t>(GetSerializeSize(TX_WITH_WITNESS(block)))};
```
Done, went with `unsigned int`

l0rinc replied:
>  for serialization fixed-size ints are required

Yes, it's the same argument as in https://github.com/bitcoin/bitcoin/pull/31490#discussion_r1908760710

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -1092,33 +1092,24 @@ bool BlockManager::ReadRawBlockFromDisk(std::vector<uint8_t>& block, const FlatF
         return FlatFilePos();
     }
-
-    // Open history file to append
```
I've reverted the comments that were used to group functionality (write header ... write [undo[block), but I've deleted the ones that were just repeating code. Especially after extracting the constant to names values, they don't add any value.

l0rinc commented:
Thanks a lot for the reviews, I've pushed [these changes](https://github.com/bitcoin/bitcoin/compare/2ff0ea366c61b2fcb80ad1f711480915c7a7aa2e..e25f0292117bce2e9f82240a4bfe86cc91e2f6c2):
* changed title and commit messages to reflect the new purpose of refactoring and code cleanup;
* removed repetitive mention of follow-up #31539;
* split benchmark renaming into a separate diff to ease review;
* replaced all temporary asserts to static_asserts (with todos to help the reviewers);
* added assertion to the scripted diff to make sure the new names don't exist yet;
* changed `SaveBlock[Undo]` replacements to `WriteBlock[Undo]` in the scripted diff;
* unified type sizes in new code to `unsigned int`;
* reverted comments that are used to group code parts.

ryanofsky reviewed (approved):
Code review ACK 223081ece651dc616ff63d9ac447eedc5c2a28fa. Since last review, "Save" was renamed to "Write", uint32_t references were dropped, some log statements and comments were improved as suggested, and a lot of tweaks made to commits and commit messages which should make this easier to review.

ryanofsky commented on src/bench/readwriteblock.cpp:
```diff
@@ -0,0 +1,68 @@
+    return block;
+}
+
+static void SaveBlockBench(benchmark::Bench& bench)
```
In commit "bench: add SaveBlockBench" (86b85bb11f8999eb59e34bd026b0791dc866f2eb)

Could rename SaveBlock to WriteBlock here too

l0rinc replied:
Right, if I need to edit, I'll rename this as well

hodlinator reviewed (approved):
ACK 223081ece651dc616ff63d9ac447eedc5c2a28fa

Thanks for reorganizing the first commits!
Confirmed that `git -c log.showSignature=false log --oneline --follow src/bench/readwriteblock.cpp` shows 7 commits.

Cool with the sanity-check in the scripted diff, not sure I've seen that before.


### Commit message dfb2f9d004860c95fc6f0d4a016a9c038d53a475

Might add some more context:
"Similarly +to `UndoWriteToDisk` in parent commit+, `WriteBlockToDisk` wasn't really extracting"


### Commit message 42bc4914658d9834a653bd1763aa8f0d54355480

(What's the inspiration for all the semicolons? Doesn't appear to be one of the uses described here: https://grammarist.com/punctuation/how-to-use-semicolons-in-a-list/)


### Benchmarked with new bench target

```
₿ build/src/bench/bench_bitcoin -filter=SaveBlockBench -min-time=10000
```

#### At second commit (86b85bb11f8999eb59e34bd026b0791dc866f2eb)

```
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        3,172,375.74 |              315.22 |    0.6% |   20,053,788.28 |    9,071,225.51 |  2.211 |   3,133,287.73 |    0.5% |     11.12 | `SaveBlockBench`
```

(Median result of 3 runs).

#### At final commit (223081ece651dc616ff63d9ac447eedc5c2a28fa)

```
|               ns/op |                op/s |    err% |          ins/op |          cyc/op |    IPC |         bra/op |   miss% |     total | benchmark
|--------------------:|--------------------:|--------:|----------------:|----------------:|-------:|---------------:|--------:|----------:|:----------
|        3,159,241.92 |              316.53 |    1.4% |   19,805,232.51 |    8,963,495.75 |  2.210 |   3,080,238.64 |    0.4% |     11.08 | `SaveBlockBench`
```

(Median result of 3 runs).

#### Conclusion

Unfortunately this confirms that serializing blocks is insanely fast, making this PR more of a refactor than an optimization.

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -963,70 +936,66 @@ bool BlockManager::FindUndoPos(BlockValidationState& state, int nFile, FlatFileP
-            if (!FlushUndoFile(_pos.nFile, true)) {
-                LogPrintLevel(BCLog::BLOCKSTORAGE, BCLog::Level::Warning, "Failed to flush undo file %05i\n", _pos.nFile);
+            if (!FlushUndoFile(pos.nFile, true)) {
+                LogPrintLevel(BCLog::BLOCKSTORAGE, BCLog::Level::Warning, "Failed to flush undo file %05i\n", pos.nFile);
```
Wouldn't `LogWarning` replace `BCLog::BLOCKSTORAGE` with `BCLog::LogFlags::ALL`?

hodlinator replied:
Correct, by "non-categorized" I was referring to `BCLog::LogFlags::ALL`.

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -963,70 +936,66 @@ bool BlockManager::FindUndoPos(BlockValidationState& state, int nFile, FlatFileP
+        // Open history file to append
+        AutoFile fileout{OpenUndoFile(pos)};
+        if (fileout.IsNull()) {
+            LogError("OpenUndoFile failed");
```
Yeah, I prefer changing it minimally when moving, to make sure I don't hide any changes there.

sedited reviewed (approved):
ACK 223081ece651dc616ff63d9ac447eedc5c2a28fa

I don't see a change in the runtime of the benchmarks, but these are some good refactors that I have wanted for some time.

andrewtoth reviewed (approved):
ACK 223081ece651dc616ff63d9ac447eedc5c2a28fa

l0rinc commented on src/node/blockstorage.cpp:
```diff
@@ -963,70 +936,66 @@ bool BlockManager::FindUndoPos(BlockValidationState& state, int nFile, FlatFileP
-        if (!FindUndoPos(state, block.nFile, _pos, ::GetSerializeSize(blockundo) + 40)) {
-            LogError("%s: FindUndoPos failed\n", __func__);
+        FlatFilePos pos;
+        const unsigned int blockundo_size{static_cast<unsigned int>(GetSerializeSize(blockundo))};
```
That's what I was going for originally, but this was explicitly requested by @ryanofsky in https://github.com/bitcoin/bitcoin/pull/31490#discussion_r1907299844

Sjors commented:
This PR renamed UndoReadFromDisk to ReadBlockUndo. Hopefully this comment makes this easier to find.

l0rinc commented:
yes, it was a leftover from the renames, suggested in https://github.com/bitcoin/bitcoin/pull/31490#pullrequestreview-2509562827

Did the rename cause any problems?

Sjors commented:
@l0rinc I was rebasing a branch that used the old name and didn't know what it was renamed too, but otherwise no problem.
